### PR TITLE
FIX: selection of wrong interface in case of argument with matching type and mismatched dimensions

### LINKED ISSUE #3244: Wrong interface selected
Requested in #3046.

```fortran
module temp_module

interface all_close

procedure :: all_close_1_csp
procedure :: all_close_2_csp
end interface all_close
contains

logical pure function all_close_1_csp(a, b) result(close)

    complex, intent(in) :: a(:), b(:)
end function all_close_1_csp
logical pure function all_close_2_csp(a, b) result(close)

    complex, intent(in) :: a(:,:), b(:,:)
end function all_close_2_csp

end module

program example_math_all_close
use temp_module, only: all_close

complex :: z(4, 4)

print *, all_close(z, z)

end program example_math_all_close
```

> LFortran

```console
% lfortran a.f90 
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2370
    LCompilers::print_stack_on_segfault();
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2336
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 949
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 360
    compiler_options, run_fn, infile);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9545
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/pass/pass_manager.h", line 302
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/pass/pass_manager.h", line 162
    _passes_db[passes[i]](al, *asr, pass_options);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/pass/pass_array_by_data.cpp", line 651
    u.visit_TranslationUnit(unit);
  File "../libasr/asr.h", line 39854
  File "../libasr/asr.h", line 5088
  File "../libasr/asr.h", line 4801
  File "../libasr/asr.h", line 39866
  File "../libasr/asr.h", line 39848
  File "../libasr/asr.h", line 5105
  File "../libasr/asr.h", line 4840
  File "../libasr/asr.h", line 40092
  File "../libasr/asr.h", line 5152
  File "../libasr/asr.h", line 4879
  File "/Users/pranavchiku/repos/lfortran/src/libasr/pass/pass_array_by_data.cpp", line 578
    visit_Call(x);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/pass/pass_array_by_data.cpp", line 540
    throw LCompilersException("Number of arguments in the new "
LCompilersException: Number of arguments in the new function call doesn't satisfy min_args <= new_args.size() <= max_args, 6 <= 10 <= 6
```

When we see the function call to `all_close`

```clojure
                        (FunctionCall
                            5 all_close@all_close_1_csp !-- should have call all_close_2_csp
                            5 all_close
                            [((ArrayPhysicalCast
                                (Var 5 z)
                                FixedSizeArray
                                DescriptorArray
                                (Array
                                    (Complex 4)
                                    [((IntegerConstant 1 (Integer 4))
                                    (IntegerConstant 4 (Integer 4)))
                                    ((IntegerConstant 1 (Integer 4))
                                    (IntegerConstant 4 (Integer 4)))]
                                    DescriptorArray
                                )
                                ()
                            ))
                            ((ArrayPhysicalCast
                                (Var 5 z)
                                FixedSizeArray
                                DescriptorArray
                                (Array
                                    (Complex 4)
                                    [((IntegerConstant 1 (Integer 4))
                                    (IntegerConstant 4 (Integer 4)))
                                    ((IntegerConstant 1 (Integer 4))
                                    (IntegerConstant 4 (Integer 4)))]
                                    DescriptorArray
                                )
                                ()
                            ))]
                            (Logical 4)
                            ()
                            ()
                        )
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #3244, towards #3046.

With this PR and https://github.com/Pranavchiku/stdlib-fortran-lang/commit/578ec4253e6f4dc0fb6618939a32c8053de753c8 we get `example_math_all_close` working.

### CODE REVIEW

**Review by czgdp1807 — COMMENTED** [review]
This is not the correct approach. Check my PR https://github.com/lfortran/lfortran/pull/3249

**Comment by Pranavchiku** [comment]
Initially I did the same, i.e. https://github.com/lfortran/lfortran/pull/3249/commits/ec5416635aaeaa827235fb1c8adcdfe1137b39c0. This leads to 
```console
[ 42%] Building Fortran object CMakeFiles/intrinsics_121.dir/intrinsics_121.f90.o
[ 42%] Built target intrinsics_118
[ 42%] Linking Fortran executable intrinsics_120
[ 42%] Building Fortran object CMakeFiles/intrinsics_123.dir/intrinsics_123.f90.o
[ 42%] Building Fortran object CMakeFiles/intrinsics_122.dir/intrinsics_122.f90.o
[ 42%] Building Fortran object CMakeFiles/intrinsics_124.dir/intrinsics_124.f90.o
[ 42%] Building Fortran object CMakeFiles/intrinsics_125.dir/intrinsics_125.f90.o
[ 42%] Building Fortran object CMakeFiles/intrinsics_126.dir/intrinsics_126.f90.o
[ 42%] Linking Fortran executable intrinsics_121
semantic error: Arguments do not match for any generic procedure, random_number
  --> /Users/pranavchiku/repos/lfortran/integration_tests/intrinsics_125.f90:15:5
   |
15 |     call random_number(array)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
make[2]: *** [CMakeFiles/intrinsics_125.dir/build.make:75: CMakeFiles/intrinsics_125.dir/intrinsics_125.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10015: CMakeFiles/intrinsics_125.dir/all] Error 2
```

Ahh, I see then you added support for arrays at https://github.com/lfortran/lfortran/pull/3249/commits/1bd8e10c29b3f7a34ae74194ebcb4507c5daf477

**Comment by czgdp1807** [comment]
Yes. More error means with a conceptually correct fix means that the new error should be fixed.

**Comment by Pranavchiku** [comment]
But I guess there will be a trouble for elemental functions / intrinsics in that approach.

**Comment by czgdp1807** [comment]
I am working on that.

**Comment by Pranavchiku** [comment]
Perfect, this is not urgent. I'll wait for your PR. Thank you!

File: src/libasr/asr_utils.h [hunk 1]
```diff
@@ -3385,29 +3385,48 @@ inline bool check_equal_type(ASR::ttype_t* x, ASR::ttype_t* y, bool check_for_di
 bool select_func_subrout(const ASR::symbol_t* proc, const Vec<ASR::call_arg_t>& args,
     Location& loc, const std::function<void (const std::string &, const Location &)> err);
 
+template <typename T>
+bool argument_types_and_dimensions_match(const Vec<ASR::call_arg_t>& args,
+        const T &sub);
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/libasr/asr_utils.h:3390]: Not needed yet.

**Comment by Pranavchiku** [comment]
Closing with reference to #3249.

**Pull request closed without merging**